compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -242,8 +242,9 @@ object Types {
           // If the type is `T | Null` or `T | Nothing`, and `T` derivesFrom the class,
           // then the OrType derivesFrom the class. Otherwise, we need to check both sides
           // derivesFrom the class.
-          if defn.isBottomType(tp.tp1) then loop(tp.tp2)
-          else loop(tp.tp1) && (defn.isBottomType(tp.tp2) || loop(tp.tp2))
+          loop(tp.tp1) && loop(tp.tp2)
+          || tp.tp1.isNullType && loop(tp.tp2)
+          || tp.tp2.isNullType && loop(tp.tp2)
```

Comment:
Ther's a typo in here (`loop(tp.tp2)` should be `loop(tp.tp1)`), and this isn't quite the same thing as before; it doesn't handle `T | Nothing`, just `T | Null`. Instead of changing this I suggest changing isBottomType / isBottomTypeAfterErasure to use `tp.isRef` instead of `tp.derivesFrom` (you can't have a subclass of Null or Nothing anyway).

Reply:
Changing `derivesFrom` to `isRef` does not work since for unions and intersections though.